### ISSUE TITLE
Canonical approach for new vectorized functions

### ISSUE THREAD

**Issue by max-sixty** [pydata/xarray#1992]
We are moving some code over from pandas to Xarray, and one of the biggest missing features is exponential functions, e.g. `series.ewm(span=20).mean()`.

It looks like we can write these as gufuncs without too much trouble in numba. But I also notice that [numbagg](https://github.com/shoyer/numbagg) hasn't changed in a while and that we chose bottleneck for many of the functions in Xarray.

- Is numba a good approach for these?
- As well as our own internal use, could we add numba functions to Xarray, or are there dependency issues?
- Tangentially, I'd be interested why we're using bottleneck rather than numbagg for the existing functions

**Comment by shoyer** [comment]
I started numbagg with the intent of using it for xarray, but never got around to finishing it up. I think it would be perfectly fine to add as an optional dependency. Mostly we are using bottleneck instead because it's already out there and is slightly more complete, but certainly writing new aggregation functions with numbagg rather than bottleneck would be much easier.

One thing I was also excited about doing with numbagg was writing fast grouped aggregation functions, e.g., to make xarray's groupby().mean() comparable in performance to pandas. I had some promising preliminary results, but now I can't find the notebook where I wrote them!

Out of curiosity, I ran the benchmark suite from the numbagg README again. Both Numba and bottleneck have been significantly improved over the past three years, but Numba now beats bottleneck across the board:
```
%timeit numbagg.nanmean(x)
1000 loops, best of 3: 1.92 ms per loop
%timeit numbagg.nanmean(x, axis=0)
100 loops, best of 3: 2.75 ms per loop
%timeit numbagg.nanmean(x, axis=1)
1000 loops, best of 3: 1.86 ms per loop

%timeit bottleneck.nanmean(x)
100 loops, best of 3: 2.75 ms per loop
%timeit bottleneck.nanmean(x, axis=0)
100 loops, best of 3: 3.47 ms per loop
%timeit bottleneck.nanmean(x, axis=1)
100 loops, best of 3: 2.72 ms per loop
```
